File: Notes/Computer Science/Algorithms/Meta/Sebastien Rubens Meta Puzzles Solutions/scala/l2_director_photography2.scala
// meta_puzzles by Sebastien Rubens
//
// Please go to https://github.com/seb-pg/meta_puzzles/README.md
// for more information
//
// To the extent possible under law, the person who associated CC0 with
// meta_puzzles has waived all copyright and related or neighboring rights
// to meta_puzzles.
//
// You should have received a copy of the CC0 legalcode along with this
// work.  If not, see <http://creativecommons.org/publicdomain/zero/1.0/>.

package l2_director_photography2

import scala.collection.mutable.ArrayBuffer

object Solution {
    def getArtisticPhotographCount(N: Int, C: String, X: Int, Y: Int): Long = {

        def _getArtisticPhotographCount(N: Int, C: String, X: Int, Y: Int): Long =
        {
            val _X = X
            val _Y = Y

            class Counts(var p: Int, var b: Int) {}

            // count the number of Ps or Bs till a position i: O(N)
            val w = _Y + 1
            val count = new Counts(0, 0)
            val counts = new ArrayBuffer[Counts](C.length + w.toInt * 2)
            for (i <- 0 until w) // add space at the beginning to avoid special treatment of indices later
                counts += new Counts(0, 0)
            for (ci <- C) {
                if (ci == 'P')
                    count.p += 1
                else if (ci == 'B')
                    count.b += 1
                counts += new Counts(count.b, count.p)
            }
            val last = counts.last
            for (i <- 0 until w) // add space at the end to avoid special treatment of indices later
                counts += last

            // To make things more readable, we are finding first the point where 'A' is found: O(N)
            val possible = new ArrayBuffer[Int](C.length)
            var j = w
            for (ci <- C) {
                if (ci == 'A')
                    possible += j
                j += 1
            }

            // Count PABs : O(N)
            val X1 = _X - 1
            val Y1 = _Y + 1
            var nb: Long = 0
            for (i <- possible) {
                val a = counts(i - _X).p - counts(i - Y1).p
                val b = counts(i + _Y).b - counts(i + X1).b
                nb += a.toLong * b.toLong
            }

            // Count BAPs : O(N)
            for (i <- possible) {
                val a = counts(i - _X).b - counts(i - Y1).b
                val b = counts(i + _Y).p - counts(i + X1).p
                nb += a.toLong * b.toLong
            }

            return nb
        }

        return _getArtisticPhotographCount(N, C, X, Y).toLong
    }

    class Args(val C: String,
               val X: Int,
               val Y: Int,
               val res: Long ) extends test.Result[Long] {
        override def get_result(): Long = res
    }

    def tests(): Int =
    {
        val wrapper = (p: Args) => getArtisticPhotographCount(p.C.length, p.C, p.X, p.Y)

        val args_list = Array[Args](
            new Args( "APABA", 1, 2, 1 ),
            new Args( "APABA", 2, 3, 0 ),
            new Args( ".PBAAP.B", 1, 3, 3 ),
            // extra1
            new Args( "PP.A.BB.B", 1, 3, 4 )
        )

        return test.TestAll.run_all_tests("l2_director_photography2", args_list, wrapper)
    }
}

File: Notes/Computer Science/Algorithms/Meta/Sebastien Rubens Meta Puzzles Solutions/cs/l2_director_photography2.cs
// meta_puzzles by Sebastien Rubens
//
// Please go to https://github.com/seb-pg/meta_puzzles/README.md
// for more information
//
// To the extent possible under law, the person who associated CC0 with
// meta_puzzles has waived all copyright and related or neighboring rights
// to meta_puzzles.
//
// You should have received a copy of the CC0 legalcode along with this
// work.  If not, see <http://creativecommons.org/publicdomain/zero/1.0/>.

using System;
using System.Collections.Generic;

namespace l2_director_photography2
{
class Solution {

    struct Counts
    {
        public int p;
        public int b;

        public Counts(int _p = 0, int _b = 0)
        {
            p = _p;
            b = _b;
        }

        public Counts(Counts other)
        {
            p = other.p;
            b = other.b;
        }
    }

    public static T _getArtisticPhotographCount<T>(int N, string C, int X, int Y) where T : unmanaged
    {
        // https://www.metacareers.com/profile/coding_puzzles/?puzzle=870874083549040
        // Constraints
        //      1 ≤ N ≤ 200         N is the number of cells in a row
        //      1 ≤ X ≤ Y ≤ N       X,Y are the distance between a photograph and an actor
        // Complexity: O(N) ~ O(N * (Y-X+1)) because Y-X << N

        // Note: the solution is not really elegant speed-wise, but focuses on the algorithm
        // The solution calculates the pairs of sub string (of width w), surrounding 'A' characters, this is O(N)
        var w = Y + 1;  // w = width of the interval
        var count = new Counts();
        var counts = new List<Counts>(C.Length + w * 2);
        for (int i = 0; i < w; ++i)
            counts.Add(new Counts());
        foreach (char ci in C)
        {
            if (ci == 'P')
                ++count.p;
            else if (ci == 'B')
                ++count.b;
            counts.Add(new Counts(count));
        }
        var last = counts[^1];
        for (int i = 0; i < w; ++i)
            counts.Add(last);

        // To make things more readable, we are finding first the point where 'A' is found: O(N)
        var possible = new List<int>(C.Length);
        int j = w;
        foreach (var ci in C)
        {
            if (ci == 'A')
                possible.Add(j);
            ++j;
        }

        // Count PABs : O(N)
        long nb = 0;
        foreach (int i in possible)
        {
            var a = counts[i - X].p - counts[i - Y - 1].p;
            var b = counts[i + Y].b - counts[i + X - 1].b;
            nb += (long)a * (long)b;
        }

        // Count BAPs : O(N)
        foreach (int i in possible)
        {
            var a = counts[i - X].b - counts[i - Y - 1].b;
            var b = counts[i + Y].p - counts[i + X - 1].p;
            nb += (long)a * (long)b;
        }

        return (T)Convert.ChangeType(nb, typeof(T));
    }

    public long getArtisticPhotographCount(int N, string C, int X, int Y) {
        // https://www.metacareers.com/profile/coding_puzzles/?puzzle=156565259776376
        // Constraints
        //      1 ≤ N ≤ 300,000     N is the number of cells in a row
        //      1 ≤ X ≤ Y ≤ N       X,Y are the distance between a photograph and an actor
        // Complexity: O(N) ~ O(N * (Y-X+1)) because Y-X << N

        return _getArtisticPhotographCount<long>(N, C, X, Y);
    }

    class Args : test_all.Res<long>
    {
        public string C;
        public int X;
        public int Y;
    }

    public static int tests()
    {
        var s = new Solution();

        Func<Args, long> wrapper = (Args p) => s.getArtisticPhotographCount(p.C.Length, p.C, p.X, p.Y);

        var args_list = new List<Args> {
            new Args { C="APABA", X=1, Y=2, res=1 },
            new Args { C="APABA", X=2, Y=3, res=0 },
            new Args { C=".PBAAP.B", X=1, Y=3, res=3 },
            // extra1
            new Args { C="PP.A.BB.B", X=1, Y=3, res=4 },
        };

        return test_all.TestAll.run_all_tests("l2_director_photography2", args_list, wrapper);
    }

}

}
